sympy/polys/specialpolys.py

```diff
@@ -149,6 +149,9 @@ def interpolating_poly(n, x, X='x', Y='y'):
     if isinstance(Y, string_types):
         Y = symbols("%s:%s" % (Y, n))
 
+    if x in X:
```

Comment:
What should be be done if `X.has(x)` or `Y.has(x)`? The result will probably not be the Lagrange polynomial in `x`.

Reply:
The way this PR behaves, the corresponding value of Y will be returned:
```
>>> interpolating_poly(3,x).subs(x,x0)
y0
>>> interpolating_poly(3,x0)
y0
```

Reply:
I was thinking of the possibility that `X` or `Y` were a list containing something like `log(x)` (and not `x` alone). The result would not be a polynomial in `x`. Maybe those cases should be checked.

Reply:
Like this?
```python
>>> X,Y = (1,2,3,log(x)),(4,5,6,7)
>>> x,y
(x, y)
>>> interpolating_poly(3,x,X,Y)==interpolating_poly(3,y,X,Y).subs(y,x)
True
```

Reply:
I mean this:
```
>>> X,Y = (1,2,3,log(x)),(4,5,6,7)
>>> p = interpolating_poly(3,x,X,Y)
>>> p
2*(x - 3)*(x - 2)*(x - log(x)) - 5*(x - 3)*(x - 1)*(x - log(x)) + 3*(x - 2)*(x - 1)*(x - log(x))
```
`p` is not a polynomial in `x`. I think that `interpolating_poly` should return a polynomial while `interpolate` could return any value.

Reply:
/!\ if you don't set n to the length of X and Y you get a "feature":

```python
>>> interpolating_poly(1,x,(x0,x1),(4,5))
4*x - 4*x1
>>> interpolating_poly(2,x,(x0,x1),(4,5)).simplify()
(-x + 5*x0 - 4*x1)/(x0 - x1)
```

As for allowing a variable point in the X and Y, I don't understand what is obtained in such cases...but I'm not sure that we shouldn't allow the user to unix-like privilege of "shooting oneself in the foot". Fixed-point iteration comes to mind as a seemingly odd rearrangement of an expression but it has a purpose.

This seems consistent:
```python
>>> solve(interpolating_poly(2,x,(y,2),(4,5))-4,x)
[y]
>>> interpolating_poly(2,x,(x,2),(4,5))
4
```
But what does this mean?
```
>>> interpolating_poly(2,x,(1/x,2),(4,5)).simplify()
(x**2 + 8*x - 5)/(2*x - 1)

>>> solve(interpolating_poly(2,x,(1/x,2),(4,5))-4)
[-1, 1]

>>> solve(interpolating_poly(2,x,(x**2,2),(4,5))-4)
[0, 1]
```

Reply:
> allowing a variable point in the X and Y,

That is what I would not like as the result need not be a polynomial in general. I think that `X` and `Y` should be checked before constructing the polynomial. I also have doubts about the `x in X` case as the result is not guaranteed to be a polynomial in `x`. So perhaps we should raise an exception if the given symbol `x` appears in any of the `X` or `Y`.

Reply:
Given that
```python
>>> X,Y = (1,2,3,log(x)),(4,5,6,7)
>>> x,y
(x, y)
>>> interpolating_poly(4,x,X,Y)==interpolating_poly(4,y,X,Y).subs(y,x)
```
are you in favor of only raising an error in `interpolating_poly`? If so, I will add a `strict` flag that will check to see that all Expr elements of X and Y have free symbols that are free of x. `interpolate` will then call with the `strict=False` flag.

Reply:
> only raising an error in interpolating_poly

Yes, I think that no restrictions is necessary in `interpolate`. `x` could be any point.

Currently, `interpolating_poly` and `interpolate` essentially do the same thing except that `interpolate` also admits a dict of data. I would like to make them two different functions, one returning a *polynomial* and the other returning its *value* at a single point. 

`interpolating_poly` would return a polynomial (expression) in the given symbol `x`. The symbol should not appear in data to make sure that the result really is a polynomial.

`interpolate` would return the value of the polynomial at the given point `x` that need not be a symbol. `interpolate` would call `interpolating_poly` with a symbol that does not appear in data (most likely a new dummy symbol) and then substitute `x` for the symbol. That way there should be no need for a `strict` flag. (`interpolating_poly` would always check the data.) `interpolate` could also handle the `x in X` case.

Reply:
That sounds like a good idea. I have incorporated these changes and also made `interpolating_poly` respect the `n` value. I also removed redundant code which appears to be less efficient (testing 20 data points indicated the removed code was about 10 times slower.